shared/js/moz_intl.js

```diff
@@ -71,71 +71,50 @@ global.mozIntl = {
         resolvedOptions.hour12 === undefined) {
       resolvedOptions.hour12 = navigator.mozHour12;
     }
 
-    if (resolvedOptions.dayperiod === undefined &&
-        resolvedOptions.hour12 === true) {
-      resolvedOptions.dayperiod = true;
-    }
 
     var intlFormat = new Intl.DateTimeFormat(locales, resolvedOptions);
 
     resolvedOptions.locale = intlFormat.resolvedOptions().locale;
     resolvedOptions.hour12 = intlFormat.resolvedOptions().hour12;
 
-    // This is needed for a workaround for bug 1208808
-    // Remove when that bug is fixed
-    var hourFormatter;
-    if (resolvedOptions.dayperiod !== undefined &&
-        resolvedOptions.hour12 === true) {
-      hourFormatter = Intl.DateTimeFormat(locales, {
-        hour: 'numeric',
-        hour12: false
-      });
-    }
-
     return {
       resolvedOptions() { return resolvedOptions; },
       format: function(date, tokenFormats) {
-        var dayPeriod;
-
-        var string = intlFormat.format(date);
-
-        if (resolvedOptions.dayperiod === false &&
-            resolvedOptions.hour12 === true) {
-          dayPeriod = getDayPeriodTokenForDate(date, hourFormatter);
-          string = string.replace(dayPeriod, '').trim();
-        } else if (resolvedOptions.dayperiod === true &&
-           options.hour === undefined) {
-          dayPeriod = getDayPeriodTokenForDate(date, hourFormatter);
-          const hour = date.toLocaleString(navigator.languages, {
-            hour12: true,
-            hour: 'numeric'
-          }).replace(dayPeriod, '').trim();
-          string = string.replace(hour, '').trim();
+
+        if (resolvedOptions.dayperiod !== false &&
+            tokenFormats === undefined) {
+          return intlFormat.format(date);
+        }
+
+        var result = intlFormat.formatWithPosition(date);
+        var string = result.value;
+
+        if (resolvedOptions.dayperiod === false) {
+          if (tokenFormats === undefined) {
+            tokenFormats = {};
+          }
+          // let's replace dayperiod token with empty string
+          tokenFormats.dayperiod = '';
         }
 
         for (var token in tokenFormats) {
-          if (token === 'dayperiod' &&
-            resolvedOptions.hour12 === false) {
+          if (!result.fieldPosition.hasOwnProperty(token)) {
             continue;
           }
-          const localOptions = {
-            [token]: resolvedOptions[token],
-          };
 
-          var formatter = global.mozIntl.DateTimeFormat(
-            navigator.languages, localOptions);
-          var tokenString = formatter.format(date);
+          var tokenPosition = result.fieldPosition[token];
+
+          //XXX: In the future we may be smarter and use position to
+          // concatenate new string, but for now, we can just replace
+          // the tokenString using str.replace
+          var tokenString =
+            string.substr(tokenPosition.beginIndex, tokenPosition.endIndex);
```

Comment:
Wouldn't it make more sense to use `reduce` on `Object.keys(result.fieldPosition)` here instead of the `for-in` and concatenate, as you suggested in the comment?